### PR TITLE
Convert some more mapbox gl styles

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR adds support for:
- converting text-offset values for labels set to line placement
- converting interpolated icon marker images
- converting interpolated fill pattern images

Interpolated icon markers GIF:
Peek 2020-09-11 14-33
_Look for the restaurant, bar, etc. markers going from round dots to larger icons._

### CODE REVIEW

File: src/core/vectortile/qgsmapboxglstyleconverter.cpp [hunk 1]
```diff
@@ -1938,24 +1992,130 @@ QImage QgsMapBoxGlStyleConverter::retrieveSprite( const QString &name, QgsMapBox
...
+                           .arg( size.width() );
+
+      QSize s;
+      for ( int i = 0; i < stops.size() - 1; ++i )
+      {
+        ;
+        sprite = retrieveSprite( stops.value( 0 ).toList().value( 1 ).toString(), context );
+        if ( !sprite.isNull() )
+        {
+          s = sprite.size() * context.pixelSizeConversionFactor();
```
Comment (nyalldawson) [inline src/core/vectortile/qgsmapboxglstyleconverter.cpp:2060]: This same code is used multiple times - could you pull it out to a single function?

File: src/core/vectortile/qgsmapboxglstyleconverter.cpp [hunk 2]
```diff
@@ -1039,6 +1029,56 @@ void QgsMapBoxGlStyleConverter::parseSymbolLayer( const QVariantMap &jsonLayer,
...
+        }
+
+        if ( !textOffset.isNull() )
+        {
+          labelSettings.distUnits = context.targetUnit();
+          labelSettings.dist = std::abs( textOffset.y() ) - textSize;
+          labelSettings.lineSettings().setPlacementFlags( textOffset.y() > 0.0 ? QgsLabeling::BelowLine : QgsLabeling::AboveLine );
+          if ( !textSizeProperty.isEmpty() && textOffsetProperty.isEmpty() )
+          {
+            ddLabelProperties.setProperty( QgsPalLayerSettings::LabelDistance, QStringLiteral( "with_variable('text_size',%2,%1*@text_size-@text_size)" ).arg( std::abs( textOffset.y() / textSize ) ).arg( ( textSizeProperty ) ) );
```
Comment (nyalldawson) [inline src/core/vectortile/qgsmapboxglstyleconverter.cpp:1073]: It is this actually needed? I can't work out the conditions where it's used

Comment (nirvn) [reply src/core/vectortile/qgsmapboxglstyleconverter.cpp:1073]: It's needed when the text size is dynamically defined via a property.

Comment (nirvn) [reply src/core/vectortile/qgsmapboxglstyleconverter.cpp:1073]: Ie, the offset might be static, but em being a derivative of text size, if the latter changes, we need to adjust.

File: src/core/vectortile/qgsmapboxglstyleconverter.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/vectortile/qgsmapboxglstyleconverter.cpp:2060]: Will do.

**Comment by nirvn** [comment]
@nyalldawson , review addressed.

**Comment by nirvn** [comment]
@nyalldawson , one more bit of styling converted, the text-rotation-alignment:viewport mode for line labels. 

That's often used (alongside background shields) for road numbers:
Screenshot from 2020-09-13 09-29-34

As for the shield itself, I don't think we can convert it ATM.

**Comment by nirvn** [comment]
Well, no, let's do shields after all:
image

:smile:

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**